aito/cli/parser.py

```
@@ -0,0 +1,144 @@
+        args.add_argument('--server', '-s', type=str, help='server to connect to', default='.env')
+        args.add_argument('--port', '-P', type=str, help='port to connect to', default='.env')
+        args.add_argument('--database', '-d', type=str, help='database to connect to', default='.env')
+        args.add_argument('--username', '-u', type=str, help='username for authentication', default='.env')
+        args.add_argument('--password', '-p', type=str, help='password for authentication', default='.env')
+
+        epilog_str = '''If no database credentials flag is given, the following environment variable are used: 
+  SQL_DRIVER, SQL_SERVER, SQL_PORT, SQL_DATABASE, SQL_USERNAME, SQL_PASSWORD
+'''
+        if not self.epilog:
+            self.epilog = epilog_str
+        else:
+            self.epilog += epilog_str
+
+    def add_csv_format_default_arguments(self):
+        self.add_argument('-d', '--delimiter', type=str, default=',', help="delimiter character(default: ',')")
+        self.add_argument('-p', '--decimal', type=str, default='.', help="decimal point character(default '.')")
+
+    def add_excel_format_default_arguments(self):
+        self.add_argument('-o', '--one-sheet', type=str, metavar='sheet-name', help='read a sheet by sheet name')
+
+
+class ParseError(Exception):
+    def __init__(self, message):
+        self.message = message
+        super().__init__(message)
+
+
+class ArgType(ABC):
+    @abstractmethod
+    def __call__(self, string):
+        pass
+
+
+class PathArgType(ArgType):
+    """Customized PathType instead of argparse.FileType to handle close file more gracefully
+
+    """
+    def __init__(self, parent_must_exist: bool = False, must_exist: bool = False):
+        self.parent_must_exist = parent_must_exist
+        self.must_exist = must_exist
+
+    def __call__(self, string) -> Path:
+        try:
+            path = Path(string)
+        except Exception:
```

Use specific errors when a certain error message is printed. By using a generic Exception, some errors that are not caused by an invalid path might go missing.

---

This is a precaution check. I don't know what would it raise

---

Then just let it raise the exception and when you know which error might arise then handle them appropriately.

---

ArgumentTypeError signals the parser which argument caused the error and display the help message